File: src/matlab/com/springbok/sgp4v/Sgp4Coordinates.m
classdef (Sealed = true) Sgp4Coordinates
% Provides transformations between coordinate systems. The
% following acronyms are used:
%   TEMED - True Equator Mean Equinox of Date (Earth Centered,
%      precession and less accurate nutation modeled)
%   TETED - True Equator True Equinox of Date (Earth Centered,
%     precession and more accurate nutation modeled)
%   CIRS - Celestial Intermediate Reference System (Earth
%     Centered, precession and more accurate nutation modeled)
%   TIRS - Terrestrial Intermediate Reference System (Earth
%     Centered and fixed, precession and more accurate nutation
%     modeled)
%   ITRS - International Terrestrial Reference System (Earth
%     Centered and fixed, precession, more accurate nutation,
%     and polar motion modeled)
%   SEZ - South–East–Zenith
% See doc/matlab/com/synterein/sgp4v/Time-Space-Coordinate-Conventions.pdf.
% 
% Author: Raymond LeClair
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  methods (Static = true)
    
    function r_ltp = gei2ltp(r_gei, sensor, dNm)
    % Computes the local tangent position vector of a satellite
    % observed by a sensor given the geocentric equatorial
    % inertial position vector at a date.
    % 
    % Parameters
    %   r_gei - Geocentric equatorial inertial position vector [er]
    %   sensor - Sensor at which position vectors apply
    %   dNm -Date number at which the position vectors occur
    %
    % Returns
    %   r_ltp - Local tangent position vector [er]
      [y, m, d, H, M, S] = datevec(dNm);
      mjd = TimeUtility.date2mjd(y, m, d, H, M, S);
      TEMED = r_gei; % Returned by SGP4
      TETED = TEMED; % Ignores difference due to less
                     % accurate nutation model
      CIRS = Sgp4Coordinates.TETEDToCIRS(mjd, TETED);
      TIRS = Sgp4Coordinates.CIRSToTIRS(mjd, CIRS);
      ITRS = TIRS; % Ignores polar motion
      ITRS_0 = sensor.R_ger();
      geodetic_latitude = sensor.varphi;
      longitude = sensor.lambda;
      SEZ = Sgp4Coordinates.ITRSToSEZ(ITRS, ITRS_0, geodetic_latitude, longitude);
      r_ltp = Coordinates.R_z(pi / 2.0) * SEZ;
      
    end % gei2ltp
    
    function CIRS = TETEDToCIRS(mjd, TETED)
    % Converts TETED to CIRS.
    % 
    % Parameters
    %   mjd - Modified Julian day
    %   TETED - Position vector in the True Equator True Equinox of
    %     Date coordinate system (Earth Centered, precession and
    %     more accurate nutation modeled)
    %
    % Returns
    %   CIRS - Position vector in the Celestial Intermediate
    %     Reference System (Earth Centered, precession and more
    %     accurate nutation modeled)
      T = (mjd - 51544.5) / 36525.0;
      epsilon = (pi / (180.0 * 3600.0)) ...
                * (-0.014506 - 4612.156534 * T - 1.3915817 * T^2 ...
                   + 0.00000044 * T^3 + 0.000029956 * T^4 + 0.0000000368 * T^5);
      CIRS = Coordinates.R_z(-epsilon) * TETED;
      
    end % TETEDToCIRS
    
    function TIRS = CIRSToTIRS(mjd, CIRS)
    % Converts CIRS to TIRS
    % 
    % Parameters
    %   mjd - Modified Julian day
    %   CIRS - Position vector in the Celestial Intermediate
    %     Reference System (Earth Centered, precession and more
    %     accurate nutation modeled)
    %
    % Returns
    %   TIRS - Position vector Terrestrial Intermediate Reference
    %     System  Earth Centered and fixed, precession and more
    %     accurate nutation modeled)
      
      T = mjd - 51544.5;
      angle = 2 * pi * (0.7790572732640 + 1.00273781191135448 * T);
      TIRS = Coordinates.R_z(angle) * CIRS;
      
    end % CIRSToTIRS
    
    function SEZ = ITRSToSEZ(ITRS, ITRS_0, geodetic_latitude, longitude)
    % Converts ITRS to SEZ.
    % 
    % Parameters
    %   ITRS - Position vector in the International Terrestrial
    %     Reference System Earth Centered and fixed, precession, more
    %     accurate nutation, and polar motion modeled)
    %   ITRS_0 - Site position vector in the International
    %     Terrestrial Reference System (Earth Centered and fixed,
    %     precession, more accurate nutation, and polar motion
    %     modeled)
    %   geodetic_latitude - Site geodetic latitude [rad]
    %   longitude - Site longitude [rad]
    %
    % Returns
    %   SEZ - Position vector in the South–East–Zenith coordinate
    %     system
      X1 = ITRS - ITRS_0;
      X2 = Coordinates.R_z(longitude) * X1;
      SEZ = Coordinates.R_y(pi / 2 - geodetic_latitude) * X2;
      
    end % ITRSToSEZ
    
    function ITRS = SEZToITRS(SEZ, ITRS_0, geodetic_latitude, longitude)
    % Converts SEZ to ITRS.
    % 
    % Parameters
    %   SEZ - Position vector in the South–East–Zenith coordinate
    %     system
    %   ITRS_0 - Site position vector in the International
    %     Terrestrial Reference System (Earth Centered and fixed,
    %     precession, more accurate nutation, and polar motion
    %     modeled)
    %   geodetic_latitude - Site geodetic latitude [rad]
    %   longitude - Site longitude [rad]
    %
    % Returns
    %   ITRS - Position vector in the International Terrestrial
    %     Reference System Earth Centered and fixed, precession, more
    %     accurate nutation, and polar motion modeled)
      X1 = Coordinates.R_y(geodetic_latitude - pi / 2.0) * SEZ;
      X2 = Coordinates.R_z(-longitude) * X1;
      ITRS = X2 + ITRS_0;
      
    end % SEZToITRS                
    
    function CIRS = TIRSToCIRS(mjd, TIRS)
    % Converts CIRS to TIRS
    % 
    % Parameters
    %   mjd - Modified Julian day
    %   TIRS - Position vector in the Terrestrial Intermediate
    %     Reference System Earth Centered and fixed, precession and
    %     more accurate nutation modeled)
    %
    % Returns
    %   CIRS - Position vector in the Celestial Intermediate
    %     Reference System (Earth Centered, precession and more
    %     accurate nutation modeled)
      T = mjd - 51544.5;
      angle = 2.0 * pi * (0.7790572732640 + 1.00273781191135448 * T);
      CIRS = Coordinates.R_z(-angle) * TIRS;
      
    end % TIRSToCIRS
    
    function TETED = CIRSToTETED(mjd, CIRS)
    % Converts TETED to CIRS.
    % 
    % Parameters
    %   mjd - Modified Julian day
    %   CIRS - Position vector in the Celestial Intermediate
    %     Reference System (Earth Centered, precession and more
    %     accurate nutation modeled)
    %
    % Returns
    %   TETED - Position vector in the True Equator True Equinox of
    %     Date coordinate system (Earth Centered, precession and more
    %     accurate nutation modeled)
      T = (mjd - 51544.5) / 36525.0;
      epsilon = (pi / (180.0 * 3600.0)) ...
                * (-0.014506 - 4612.156534 * T - 1.3915817 * T^2 ...
                   + 0.00000044 * T^3 + 0.000029956 * T^4 + 0.0000000368 * T^5);
      TETED = Coordinates.R_z(epsilon) * CIRS;
      
    end % CIRSToTETED
    
    function r_gei = ltp2gei(r_ltp, sensor, dNm)
    % Computes the inertial geocentric equatorial position
    % vector of a satellite given the local tangent position
    % vector observed at a date by a given sensor.
    % 
    % Parameters
    %   r_ltp - Local tangent position vector [er]
    %   sensor - Sensor at which position vectors apply
    %   dNm - Date number at which the position vectors occur
    %
    % Returns
    %   r_gei - Inertial geocentric equatorial position vector [er]
      [y, m, d, H, M, S] = datevec(dNm);
      mjd = TimeUtility.date2mjd(y, m, d, H, M, S);
      SEZ = Coordinates.R_z(-pi / 2.0) * r_ltp;
      ITRS_0 = sensor.R_ger();
      geodetic_latitude = sensor.varphi;
      longitude = sensor.lambda;
      ITRS = Sgp4Coordinates.SEZToITRS(SEZ, ITRS_0, geodetic_latitude, longitude);
      TIRS = ITRS; % Ignores polar motion
      CIRS = Sgp4Coordinates.TIRSToCIRS(mjd, TIRS);
      TETED = Sgp4Coordinates.CIRSToTETED(mjd, CIRS);
      TEMED = TETED; % Ignores difference due to less
                     % accurate nutation model
      r_gei = TEMED; % Returned by SGP4
      
    end % ltp2gei
    
  end % methods (Static = true)
  
end % classdef

File: src/java/com/springbok/sgp4v/Sgp4Coordinates.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.sgp4v;

import java.io.Serializable;

import Jama.Matrix;

import com.springbok.station.EarthStation;
import com.springbok.twobody.ModJulianDate;

/**
 * Provides transformations between coordinate systems using methods provided by
 * Arthur Lue. The following acronyms are used:
 * <ul>
 * <li>TEMED - True Equator Mean Equinox of Date (Earth Centered, precession and
 * less accurate nutation modeled)</li>
 * <li>TETED - True Equator True Equinox of Date (Earth Centered, precession and
 * more accurate nutation modeled)</li>
 * <li>CIRS - Celestial Intermediate Reference System (Earth Centered,
 * precession and more accurate nutation modeled)</li>
 * <li>TIRS - Terrestrial Intermediate Reference System (Earth Centered and
 * fixed, precession and more accurate nutation modeled)</li>
 * <li>ITRS - International Terrestrial Reference System (Earth Centered and
 * fixed, precession, more accurate nutation, and polar motion modeled)</li>
 * <li>SEZ - South–East–Zenith</li>
 * </ul>
 * See doc/java/com/synterein/sgp4v/Time-Space-Coordinate-Conventions.pdf.
 * 
 * @author Raymond LeClair
 */
@SuppressWarnings("serial")
public final class Sgp4Coordinates implements Serializable {

	/**
	 * Computes the local tangent position vector of a satellite observed by a
	 * sensor given the geocentric equatorial inertial position vector at a
	 * date.
	 * 
	 * @param r_gei
	 *            Geocentric equatorial inertial position vector [er]
	 * @param sensor
	 *            Sensor at which position vectors apply
	 * @param dNm
	 *            Modified Julian date at which the position vectors occur
	 * @return Local tangent position vector [er]
	 */
	public static Matrix gei2ltp(Matrix r_gei, EarthStation sensor, ModJulianDate dNm) {
		Matrix TEMED = r_gei; // Returned by SGP4
		Matrix TETED = TEMED; // Ignores difference due to less accurate
								// nutation model
		Matrix CIRS = TETEDToCIRS(dNm, TETED);
		Matrix TIRS = CIRSToTIRS(dNm, CIRS);
		Matrix ITRS = TIRS; // Ignores polar motion
		Matrix ITRS_0 = sensor.get_R_ger();
		double geodetic_latitude = sensor.get_varphi();
		double longitude = sensor.get_lambda();
		Matrix SEZ = ITRSToSEZ(ITRS, ITRS_0, geodetic_latitude, longitude);
		Matrix r_ltp = com.springbok.twobody.Coordinates.R_z(Math.PI / 2.0).times(SEZ);
		// System.out.println("r_ltp"); r_ltp.print(16, 8);
		return r_ltp;
	}

	/**
	 * Converts TETED to CIRS.
	 * 
	 * @param dNm
	 *            Modified Julian day
	 * @param TETED
	 *            Position vector in the True Equator True Equinox of Date
	 *            coordinate system (Earth Centered, precession and more
	 *            accurate nutation modeled)
	 * @return Position vector in the (CIRS) Celestial Intermediate Reference
	 *         System (Earth Centered, precession and more accurate nutation
	 *         modeled)
	 */
	public static Matrix TETEDToCIRS(ModJulianDate dNm, Matrix TETED) {
		double T = (dNm.getAsDouble() - 51544.5) / 36525.0;
		double epsilon = (Math.PI / (180.0 * 3600.0)) * (-0.014506 - 4612.156534 * T - 1.3915817 * Math.pow(T, 2)
				+ 0.00000044 * Math.pow(T, 3) + 0.000029956 * Math.pow(T, 4) + 0.0000000368 * Math.pow(T, 5));
		Matrix CIRS = com.springbok.twobody.Coordinates.R_z(-epsilon).times(TETED);
		return CIRS;
	}

	/**
	 * Converts CIRS to TIRS
	 * 
	 * @param dNm
	 *            Modified Julian day
	 * @param CIRS
	 *            Position vector in the Celestial Intermediate Reference System
	 *            (Earth Centered, precession and more accurate nutation
	 *            modeled)
	 * @return Position vector in the (TIRS) Terrestrial Intermediate Reference
	 *         System (Earth Centered and fixed, precession and more accurate
	 *         nutation modeled)
	 */
	public static Matrix CIRSToTIRS(ModJulianDate dNm, Matrix CIRS) {
		double T = dNm.getAsDouble() - 51544.5;
		double angle = 2 * Math.PI * (0.7790572732640 + 1.00273781191135448 * T);
		Matrix TIRS = com.springbok.twobody.Coordinates.R_z(angle).times(CIRS);
		return TIRS;
	}

	/**
	 * Converts ITRS to SEZ.
	 * 
	 * @param ITRS
	 *            Position vector in the International Terrestrial Reference
	 *            System (Earth Centered and fixed, precession, more accurate
	 *            nutation, and polar motion modeled)
	 * @param ITRS_0
	 *            Site position vector in the International Terrestrial
	 *            Reference System (Earth Centered and fixed, precession, more
	 *            accurate nutation, and polar motion modeled)
	 * @return Position vector in the (SEZ) Site South–East–Zenith coordinate
	 *         system
	 */
	public static Matrix ITRSToSEZ(Matrix ITRS, Matrix ITRS_0, double geodetic_latitude, double longitude) {
		Matrix X1 = ITRS.minus(ITRS_0);
		Matrix X2 = com.springbok.twobody.Coordinates.R_z(longitude).times(X1);
		Matrix SEZ = com.springbok.twobody.Coordinates.R_y(Math.PI / 2 - geodetic_latitude).times(X2);
		return SEZ;
	}

	/**
	 * Converts SEZ to ITRS.
	 * 
	 * @param SEZ
	 *            Position vector in the South–East–Zenith coordinate system
	 * @param ITRS_0
	 *            Site position vector in the International Terrestrial
	 *            Reference System (Earth Centered and fixed, precession, more
	 *            accurate nutation, and polar motion modeled)
	 * @return Position vector in the (ITRS) International Terrestrial Reference
	 *         System (Earth Centered and fixed, precession, more accurate
	 *         nutation, and polar motion modeled)
	 */
	public static Matrix SEZToITRS(Matrix SEZ, Matrix ITRS_0, double geodetic_latitude, double longitude) {
		Matrix X1 = com.springbok.twobody.Coordinates.R_y(geodetic_latitude - Math.PI / 2.0).times(SEZ);
		Matrix X2 = com.springbok.twobody.Coordinates.R_z(-longitude).times(X1);
		Matrix ITRS = X2.plus(ITRS_0);
		return ITRS;
	}

	/**
	 * Converts CIRS to TIRS
	 * 
	 * @param dNm
	 *            Modified Julian day
	 * @param TIRS
	 *            Position vector in the Terrestrial Intermediate Reference
	 *            System (Earth Centered and fixed, precession and more accurate
	 *            nutation modeled)
	 * @return Position vector in the (CIRS) Celestial Intermediate Reference
	 *         System (Earth Centered, precession and more accurate nutation
	 *         modeled)
	 */
	public static Matrix TIRSToCIRS(ModJulianDate dNm, Matrix TIRS) {
		double T = dNm.getAsDouble() - 51544.5;
		double angle = 2.0 * Math.PI * (0.7790572732640 + 1.00273781191135448 * T);
		Matrix CIRS = com.springbok.twobody.Coordinates.R_z(-angle).times(TIRS);
		return CIRS;
	}

	/**
	 * Converts TETED to CIRS.
	 * 
	 * @param dNm
	 *            Modified Julian day
	 * @param CIRS
	 *            Position vector in the Celestial Intermediate Reference System
	 *            (Earth Centered, precession and more accurate nutation
	 *            modeled)
	 * @return Position vector in the (TETED) True Equator True Equinox of Date
	 *         coordinate system (Earth Centered, precession and more accurate
	 *         nutation modeled)
	 */
	public static Matrix CIRSToTETED(ModJulianDate dNm, Matrix CIRS) {
		double T = (dNm.getAsDouble() - 51544.5) / 36525.0;
		double epsilon = (Math.PI / (180.0 * 3600.0)) * (-0.014506 - 4612.156534 * T - 1.3915817 * Math.pow(T, 2)
				+ 0.00000044 * Math.pow(T, 3) + 0.000029956 * Math.pow(T, 4) + 0.0000000368 * Math.pow(T, 5));
		Matrix TETED = com.springbok.twobody.Coordinates.R_z(epsilon).times(CIRS);
		return TETED;
	}

	/**
	 * Computes the inertial geocentric equatorial position vector of a
	 * satellite given the local tangent position vector observed at a date by a
	 * given sensor.
	 * 
	 * @param r_ltp
	 *            Local tangent position vector [er]
	 * @param sensor
	 *            Sensor at which position vectors apply
	 * @param dNm
	 *            Modified Julian date at which the position vectors occur
	 * @return Inertial geocentric equatorial position vector [er]
	 */
	public static Matrix ltp2gei(Matrix r_ltp, EarthStation sensor, ModJulianDate dNm) {
		Matrix SEZ = com.springbok.twobody.Coordinates.R_z(-Math.PI / 2.0).times(r_ltp);
		Matrix ITRS_0 = sensor.get_R_ger();
		double geodetic_latitude = sensor.get_varphi();
		double longitude = sensor.get_lambda();
		Matrix ITRS = SEZToITRS(SEZ, ITRS_0, geodetic_latitude, longitude);
		Matrix TIRS = ITRS; // Ignores polar motion
		Matrix CIRS = TIRSToCIRS(dNm, TIRS);
		Matrix TETED = CIRSToTETED(dNm, CIRS);
		Matrix TEMED = TETED; // Ignores difference due to less accurate
								// nutation model
		Matrix r_gei = TEMED; // Returned by SGP4
		// System.out.println("r_gei"); r_gei.print(16, 8);
		return r_gei;
	}
}
